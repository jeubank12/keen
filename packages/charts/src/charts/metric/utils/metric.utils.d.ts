export declare type MetricType = 'percent' | 'difference' | 'compare';
declare type Options = {
  data: Readonly<Record<string, any>[]>;
  keys: string[];
  labelSelector: string;
  type: MetricType;
};
export declare type Difference = {
  value: number;
  status?: 'increase' | 'decrease' | 'static';
};
export declare const calculatePercentDifference: (
  previousValue: number,
  currentValue: number
) => number;
export declare const calculateDifference: (
  previousValue: number,
  currentValue: number,
  type: 'percent' | 'difference' | 'compare'
) => number;
export declare const setStatus: (
  previousValue: number,
  currentValue: number
) => 'static' | 'increase' | 'decrease';
export declare const generateMetric: ({
  type,
  keys,
  data,
}: Options) => {
  value: number;
  difference?: Difference;
};
export {};
