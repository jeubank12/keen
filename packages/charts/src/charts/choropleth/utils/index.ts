import { fetchMapTopology, MapType } from './map.utils';
import { generateChoropleth } from './chart.utils';

export { fetchMapTopology, generateChoropleth, MapType };
