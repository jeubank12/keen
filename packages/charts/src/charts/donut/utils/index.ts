import {
  generateDonutChart,
  LabelsPosition,
  Options as DonutChartOptions,
} from './chart.utils';

export { generateDonutChart, LabelsPosition, DonutChartOptions };
