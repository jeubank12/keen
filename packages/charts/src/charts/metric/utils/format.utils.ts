import { format } from 'd3-format';

const formatter = format('.2s');

export const formatNumber = (value: number) => formatter(value);
