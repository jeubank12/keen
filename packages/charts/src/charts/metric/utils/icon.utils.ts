import { transparentize } from 'polished';

import { MAX_RADIUS, RADIUS_DIFFERENCE } from '../constants';

type IconCircle = {
  radius: number;
  opacity: number;
  color: string;
};

const solidOpacities = [0.01, 0.01, 0.02, 0.02, 0.03, 0.03, 0.04];

const regularOpacities = [
  [0.01, 1],
  [0.02, 0.95],
  [0.03, 0.9],
  [0.04, 0.85],
  [0.05, 0.8],
  [0.08, 0.75],
  [0.01, 0.7],
];

export const generateCircles = ({
  color,
  circleStyle,
}: {
  color: string;
  circleStyle: 'solid' | 'regular';
}): IconCircle[] =>
  circleStyle === 'solid'
    ? solidOpacities.map((opacity, idx) => ({
        opacity,
        color,
        radius: MAX_RADIUS - idx * RADIUS_DIFFERENCE,
      }))
    : regularOpacities.map(([fillOpacity, colorOpacity], idx) => ({
        opacity: fillOpacity,
        color: transparentize(colorOpacity, color),
        radius: MAX_RADIUS - idx * RADIUS_DIFFERENCE,
      }));
