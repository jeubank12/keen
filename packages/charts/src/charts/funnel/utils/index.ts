import { generateFunnel, calculateStepPoints } from './chart.utils';

export { generateFunnel, calculateStepPoints };
