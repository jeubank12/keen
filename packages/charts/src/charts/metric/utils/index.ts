import { generateMetric } from './metric.utils';
import { formatNumber } from './format.utils';

export { generateMetric, formatNumber };
