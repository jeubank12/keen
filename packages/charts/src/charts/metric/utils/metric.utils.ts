import { getFromPath } from '../../../utils/selectors.utils';

export type MetricType = 'difference' | 'comparison' | 'simple';

type Options = {
  data: Readonly<Record<string, any>[]>;
  keys: string[];
  labelSelector: string;
  type: MetricType;
  usePercentDifference?: boolean;
};

export type Difference = {
  value: number;
  status?: 'increase' | 'decrease' | 'static';
};

export const calculatePercentDifference = (
  previousValue: number,
  currentValue: number
) => (currentValue / previousValue) * 100 - 100;

export const calculateDifference = (
  previousValue: number,
  currentValue: number,
  usePercentage: boolean
) => {
  switch (true) {
    case usePercentage === true:
      const value = calculatePercentDifference(previousValue, currentValue);
      return value % 1 === 0 ? value : Number(value.toFixed(2));
    case usePercentage === false:
      return previousValue - currentValue;
    default:
      return previousValue;
  }
};

export const setStatus = (previousValue: number, currentValue: number) => {
  if (previousValue === currentValue) return 'static';
  if (currentValue > previousValue) return 'increase';
  return 'decrease';
};

export const generateMetric = ({
  type,
  keys,
  data,
  usePercentDifference,
}: Options): {
  value: number;
  difference?: Difference;
} => {
  const [keyName] = keys;
  const seriesLength = data.length;
  const hasMultipleResults = seriesLength > 1;

  if (hasMultipleResults && type !== 'simple') {
    const previousValue = getFromPath(data, [seriesLength - 2, keyName]);
    const currentValue = getFromPath(data, [seriesLength - 1, keyName]);
    const usePercentage = usePercentDifference && type === 'difference';

    const valueDifference = calculateDifference(
      previousValue,
      currentValue,
      usePercentage
    );

    return {
      value: currentValue,
      difference: {
        value: Math.abs(valueDifference),
        status: setStatus(previousValue, currentValue),
      },
    };
  }

  const [firstMetric] = data;

  return {
    value: firstMetric[keyName],
  };
};
