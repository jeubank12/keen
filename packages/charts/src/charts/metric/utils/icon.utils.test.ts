import { generateCircles } from './icon.utils';

describe('@keen.io/charts - <MetricChart /> utils', () => {
  describe('generateCircles()', () => {
    it('should set color for "solid" circle style', () => {
      const color = 'black';
      const circles = generateCircles({ circleStyle: 'solid', color });

      expect(circles).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            color,
          }),
        ])
      );
    });

    it('should add color opacity for "regular" style', () => {
      const color = 'black';
      const circles = generateCircles({ circleStyle: 'regular', color });

      expect(circles).toMatchSnapshot();
    });
  });
});
