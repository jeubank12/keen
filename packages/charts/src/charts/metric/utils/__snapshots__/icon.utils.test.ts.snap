// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@keen.io/charts - <MetricChart /> utils generateCircles() should add color opacity for "regular" style 1`] = `
Array [
  Object {
    "color": "rgba(0,0,0,0)",
    "opacity": 0.01,
    "radius": 225,
  },
  Object {
    "color": "rgba(0,0,0,0.05)",
    "opacity": 0.02,
    "radius": 200,
  },
  Object {
    "color": "rgba(0,0,0,0.1)",
    "opacity": 0.03,
    "radius": 175,
  },
  Object {
    "color": "rgba(0,0,0,0.15)",
    "opacity": 0.04,
    "radius": 150,
  },
  Object {
    "color": "rgba(0,0,0,0.2)",
    "opacity": 0.05,
    "radius": 125,
  },
  Object {
    "color": "rgba(0,0,0,0.25)",
    "opacity": 0.08,
    "radius": 100,
  },
  Object {
    "color": "rgba(0,0,0,0.3)",
    "opacity": 0.01,
    "radius": 75,
  },
]
`;
